docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+sensitive to ordering outside of the static dependencies that are syntactically
+apparent in the expression.  In particular, field inference cannot change the
+outcome of initializer inference.
+
+
+A meta-goal is to keep this inference relatively lightweight and efficient to
+implement.
+
+
+In all cases, if initializer inference fails, it is an error and produces no
+result.  Implementations are free to treat this as a result of dynamic after
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Immediately evident expressions
+
+We define a set of expressions with immediately-evident types (called
+immediately-evident expressions, abbreviated IE) as follows.  Immediately
+evident expressions have an inferred type as specified below.  Any expression
+which is not an IE has no inferred type.  For every immediately evident
+expression, we also define the set of variable dependencies upon which its
+inference relies.
+
+
+* null, boolean, numeric, string, and symbol literals have their corresponding
+   type.
+   * No inference dependencies
+* A list literal or a map literal with explicit type arguments has the type
+   `List<T>` or `Map<K, V>` respectively, where `<T>` or `<K, V>` are the
+   provided type arguments.
+   * No inference dependencies
+* A non-empty list literal with no type arguments, and for which all of the
+   elements are IEs, has the type `List<T>` where `T` is the least upper bound
+   of the inferred types of the elements.
+   * If the elements do not all have an inferred type, it is an error
+   * The inference dependencies of the list literal are the collected
+      dependencies of the elements.
+* An non-const empty list literal with no type arguments has the type
+  `List<dynamic>`.
+   * No inference dependencies.
+* A const empty list literal with no type arguments has the type
+  `List<bottom>`.
+   * No inference dependencies.
+* A map literal with no type arguments, and for which all of the keys and
+   values are IEs, has the type `Map<K, V>` where `K` is the least upper bound of
+   the inferred types of the provided keys, and `V` is the least upper bound of
+   the provided values.
+   * If the keys and values do not all have an inferred type, it is an error.
+   * The inference dependencies of the map literal are the collected
+      dependencies of the keys and values.
+* An non-const empty map literal with no type arguments has the type
+  `Map<dynamic, dynamic>`.
+   * No inference dependencies.
+* A const empty map literal with no type arguments has the type `Map<bottom,
+  bottom>`.
+   * No inference dependencies.
+* A type literal has type `Type`
+   * No inference dependencies
+* A function literal with an expression body for which all parameter types are
+   type annotated and for which the return expression is an IE and has an
+   inferred type, has the corresponding function type.
```

Comment:
With 'is an IE and has an inferred type', it sounds like an IE without an inferred type can exist. I thought IE was a semantic concept (so we can say that they all have an inferred type), rather than a syntactic one (where we would specify a larger set of expressions and then filter out the ones that don't have an inferred type).

With the semantic approach, I think we shouldn't say 'it is an error' here and there, we should say 'the expression is not an IE'. With the syntactic approach, we'd need to say 'an IE which has an inferred type' in a lot of places.